docs/data/data-grid/server-side/datasource.md

```diff
@@ -0,0 +1,187 @@
+  };
+}
+```
+
+### Updating data
+
+If provided, the method `dataSource.updateRow` will be called with the `GridRowModel` object whenever the user edits a row. This method is optional and you can skip it if you don't need to update the data on the server. It will work in a similar way as the `processRowUpdate` prop.
+
+### Data grid props
+
+These data grid props will work with the server-side data source:
+
+- `dataSource: DataSource`: the data source object that you need to implement
+- `queryClient: QueryClient`: optional: if not provided, a new instance of `QueryClient` will be created internally
+- `rows`: will be ignored, could be skipped when `dataSource` is provided
+- `rowCount`: will be used to identify the total number of rows in the grid, if not provided, the grid will use the _GetRowsResponse.rowCount_ value
+
+Props related to grouped data (`treeData` and `rowGrouping`):
+
+- `getGroupKey(row: GridRowModel): string`
+
+  will be used by the grid to group rows by their parent group
+  This effectively replaces `getTreeDataPath`.
+  Consider this structure:
+
+  ```
+  - (1) Sarah // groupKey 'Sarah'
+    - (2) Thomas // groupKey 'Thomas'
+  ```
+
+  When (2) is expanded, the `getRows` function will be called with group keys `['Sarah', 'Thomas']`.
+
+- `hasChildren?(row: GridRowModel): boolean`
+
+  Will be used by the grid to determine if a row has children on server
+
+- `getChildrenCount?: (row: GridRowModel) => number`
+
+  Will be used by the grid to determine the number of children of a row on server
+
+### Existing server-side features
+
+The server-side `dataSource` will change a bit the way existing server-side features work. Whenever a valid data source is passed the features `filtering`, `sorting`, `pagination` will automatically be set to `server`.
+
+However, if for some reason, you want to explicitly work one or more of these features to work on `client`, you can do so by passing the `client` value to the feature. Here's an example:
+
+```tsx
+<DataGridPro
+  columns={columns}
+  dataSource={myCustomDataSource} // will set `filtering`, `sorting`, `pagination` to `server`
+  filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change
+/>
+```
+
+### Caching
+
+The data grid will cache the data it receives from the server. This means that if the user navigates to a page that has already been fetched, the grid will not call the `getRows` function again. This is to avoid unnecessary calls to the server.
+
+By default, the data grid will use the tanstack react-query library to cache the data. A `QueryClient` is created internally and used with the data source. You can also pass your own instance of `QueryClient` in order to have access to the cached data outside the data grid. You can also use this client to invalidate the cache when you need to.
+
+In order to access the current keys for the `QueryClient`, you can wrap the application into the `DataSourceProvider`.
```

Comment:
Example: https://codesandbox.io/s/datagridpro-basic-forked-49zpsy?file=/src/App.tsx

Reply:
Nice, do you think you could use the quick filtering in this demo instead?

Reply:
I think yes, but listening to the `onFilterModelUpdate` and catching the `filterModel.quickFilterValues` in the handler.

Reply:
Just for reference, if we don't use tanstack query or some other external library, cache sharing should be possible with our internal cache too with the same basic principle. Instead of the cache available in the user's scope, the cache could be made accessible by `apiRef`. (like `apiRef.current.serverCache` for example)